src/uart_arduino_nano.cpp

```diff
@@ -3,33 +3,112 @@
 #ifdef PLATFORM_ARDUINO_NANO
 
 #include "uart.h"
-#include <Arduino.h>
+#include <avr/io.h>
+#include <avr/interrupt.h>
 #include <stdarg.h>
+#include "ports.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+typedef struct
+{
+    volatile uint8_t busy;
+    volatile uint8_t numBytesToTransmit;
+    volatile uint8_t start;
+    volatile uint8_t end;
+    volatile uint8_t buffer[UART_BUFFER_SIZE];
+} OutputBuffer_t;
+
+static OutputBuffer_t outputBuffer = {
+    .busy = 0,
+    .numBytesToTransmit = 0,
+    .start = 0,
+    .end = 0,
+};
+
+static uint8_t isOutBufferFull()
+{
+    return (outputBuffer.end + 1) % UART_BUFFER_SIZE == outputBuffer.start;
+}
+
+static uint8_t isEmpty()
+{
+    return outputBuffer.end == outputBuffer.start;
+}
+
+static void putCharInOutBuffer(char c)
+{
+    outputBuffer.end = (outputBuffer.end + 1) % UART_BUFFER_SIZE;
+    outputBuffer.buffer[outputBuffer.end] = c;
+}
+
+static char getFromOutBuffer()
+{
+    char c = outputBuffer.buffer[outputBuffer.start];
+    outputBuffer.start = (outputBuffer.start + 1) % UART_BUFFER_SIZE;
+    return c;
+}
+
+static void sendFromOutBuffer()
+{
+    if (outputBuffer.numBytesToTransmit > 0)
+    {
+        if (!isEmpty()) UDR0 = getFromOutBuffer();
+    }
+    else
+    {
+        UCSR0B &= ~(1 << UDRE0);
```

Comment:
przydałyby się komentarze co te rejestry robią. Za miesiąc na 100% nie będziesz tego pamiętał.

Reply:
Zrobiłem define na nazwy rejestrów - nazwy zrobiły się bardziej opisowe, więc komentarze wrzuciłem tylko tam gdzie moim zdaniem było coś niejasne. Przyda się to też podczas portowania dla Atmega32 - tam rejestry mają trochę inne nazwy (np. zamiast TXEN0 jest TXEN), ale sposób konfiguracji jest identyczny.